### ISSUE TITLE
Add a way to request/show window menu

### ISSUE THREAD

**Issue by kchibisov** [rust-windowing/winit#3108]
Some systems like Wayland/Windows provide a way to ask compositor to show window menu, users can't control the content of such menu, and usually they contain compositor specific actions, like drag/maximize/etc.

Winit already shows the menu on Wayland, but you need a titlebar for that, however, when the titlebar is hidden, users can't request a menu, for example, when they draw title bar on their own.

The api should probably be like that

```
/// The system is free to ignore the request.
Window::show_window_menu(&self);
```

**Comment by daxpedda** [comment]
I removed the Wasm label, as this isn't a think on Web.

**Comment by kchibisov** [comment]
@daxpedda out of curiosity, it's done by the browser, right? 

Also, how the web apps changing right click popup works? For example, what youtube does when you right click on the video, it does show its custom popup, but how it's done?

**Comment by daxpedda** [comment]
Oh, I thought you were talking about the window menu **bar**, but you were just talking about the context menu. Sorry, not sure what the right terms here are.

> @daxpedda out of curiosity, it's done by the browser, right? And users can't ask browser to show browser default right click menu.

By default the default browser context menu is always shown on right-click. Currently we just suppress the default context menu, controlling  that suppression wouldn't be a problem at all.

> Also, how the web apps changing right click popup works? For example, what youtube does when you right click on the video, it does show its custom popup, but how it's done?

There are no custom context menus in that sense, there is only the default one and you have no control over it.
E.g. what YouTube does is just suppress the default one and add an element with CSS styling and stuff on right-click.

**Comment by kchibisov** [comment]
Ah, yeah, we need this right click menu thing, so it seems like it's applicable for web?

**Comment by daxpedda** [comment]
What is actually the intention of this API exactly? Web can just *allow* the context menu to be shown when using right-click. It can't actually open it on demand.

**Comment by kchibisov** [comment]
Ah, ok. Then it's different on the web.

The intent is to show a compositor menu when you click on decorations. So users can pick compositor action.

It looks like

![1](https://github.com/rust-windowing/winit/assets/27620401/c35e482c-fcdb-40c9-8494-61b38802b640)

**Comment by daxpedda** [comment]
So you aren't talking about just a regular context menu, but specifically the context menu related to decorations/the title bar?

Why did you talk about the right-click menu here: https://github.com/rust-windowing/winit/issues/3108#issuecomment-1727865169; was that unrelated?

**Comment by kchibisov** [comment]
Also, we already do that, if you're on Wayland and use client side decorations, you should be able to right click on winit decorations to get them. The point is to allow users to do that as well.

@daxpedda you can ask for it at any place in your window, it's usually bound to right click, but could be to any action on Wayland. You can even show them outside the decorations, since in the end, it's a "Ask system for a specific menu users can click".

**Comment by daxpedda** [comment]
Yeah, as I said, Web doesn't support opening a context menu on demand.

I wanted to know specifically about what kind of context menu we are talking about. Because the one that you seem to be talking about is specific to decorations. Because that doesn't make sense for Web at all.

**Comment by kchibisov** [comment]
Wayland done.
